File: backend/src/main/java/com/forge/report/dto/ReportRange.java
package com.forge.report.dto;

import com.forge.habit.entity.Habit;
import java.time.LocalDate;
import java.util.List;

/** Query-param-bindable time window for every report endpoint. */
public enum ReportRange {
    WEEK,
    MONTH,
    YEAR,
    ALL;

    /** Unclamped start date — see ReportService#resolveStart for the 365-day safety cap applied
     * on top of this (an ALL-time user can have habits going back years). */
    public LocalDate startDate(LocalDate today, List<Habit> habits) {
        return switch (this) {
            case WEEK -> today.minusDays(6);
            case MONTH -> today.minusDays(29);
            case YEAR -> today.minusDays(364);
            case ALL -> habits.stream().map(Habit::getStartDate).min(LocalDate::compareTo).orElse(today);
        };
    }
}

File: mobile/lib/features/reports/domain/entities/report_range.dart
enum ReportRange {
  week,
  month,
  year,
  all;

  String get label {
    switch (this) {
      case ReportRange.week:
        return 'Week';
      case ReportRange.month:
        return 'Month';
      case ReportRange.year:
        return 'Year';
      case ReportRange.all:
        return 'All time';
    }
  }

  /// Matches the backend's `ReportRange` enum constant names exactly.
  String get apiValue => name.toUpperCase();
}
